Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -345,28 +437,120 @@ void CGContextSynchronize(CGContextRef context) {
+
+    deviceContext->SetTarget(commandList.Get());
+
+    // Copy the current layer's state to the new layer.
+    auto& oldLayer = _layerStack.top();
+
+    _layerStack.emplace(commandList.Get(), rect, oldLayer);
+    auto& newLayer = _layerStack.top(); // C++17 makes .emplace return a stack<T>::reference; alas, we are C++14.
+
+    // These properties are not to be preserved across transparency layers.
+    // https://developer.apple.com/library/content/documentation/GraphicsImaging/Conceptual/drawingwithquartz2d/dq_trans_layers/dq_trans_layers.html#//apple_ref/doc/uid/TP30001066-CH210-SW1
+    // Doing so would cause:
+    // * Double shadows
+    // * Double transparency
+    // * Double blending
+    // * Double clipping
+    auto& newDrawingState = newLayer.CurrentGState();
+    newDrawingState.alpha = 1.0;
+    newDrawingState.shadowColor = { 0.f, 0.f, 0.f, 0.f };
+    // newDrawingState.blendMode = kCGBlendModeNormal; // TODO GH#1389
+    newDrawingState.clippingGeometry = nullptr;
+    if (rect) {
+        CGRect transformedClippingRegion = CGContextConvertRectToDeviceSpace(this, *rect);
+
+        if (gStateForComposingLayer.HasShadow()) {
+            // The reference platform grows the layer's clipping region by the shadow's offset and its approximate width.
+            CGFloat xdelta = fabsf(gStateForComposingLayer.shadowOffset.width);
+            xdelta += gStateForComposingLayer.shadowBlur + 1;
+
+            CGFloat ydelta = fabsf(gStateForComposingLayer.shadowOffset.height);
+            ydelta += gStateForComposingLayer.shadowBlur + 1;
+
+            transformedClippingRegion = CGRectInset(transformedClippingRegion, -xdelta, -ydelta);
+        }
+
+        ComPtr<ID2D1RectangleGeometry> rectGeometry;
+        RETURN_IF_FAILED(Factory()->CreateRectangleGeometry(__CGRectToD2D_F(transformedClippingRegion), &rectGeometry));
+        RETURN_IF_FAILED(newDrawingState.IntersectClippingGeometry(rectGeometry.Get(), kCGPathEOFill));
+    }
+
+    // A GState's global alpha is used when per-primitive drawing is not an option (such as when
+    // composing down a full transparency layer.)
+    gStateForComposingLayer.globalAlpha = gStateForComposingLayer.alpha;
+    gStateForComposingLayer.alpha = 1.0;
+    return S_OK;
+}
+
+HRESULT __CGContext::PopLayer() {
+    if (_layerStack.size() <= 1) {
+        TraceError(TAG, L"Invalid attempt to pop base layer.");
+        return E_BOUNDS;
+    }
+
+    auto outgoingLayer = std::move(_layerStack.top());
+
+    ComPtr<ID2D1Image> outgoingImageTarget;
+    RETURN_IF_FAILED(outgoingLayer.GetTarget(&outgoingImageTarget));
+
+    ComPtr<ID2D1CommandList> outgoingCommandList;
+    if (FAILED(outgoingImageTarget.As(&outgoingCommandList))) {
+        TraceError(TAG, L"Popped a layer that was NOT a command list. PANIC!");
+        return E_UNEXPECTED;
+    }
+
+    RETURN_IF_FAILED(outgoingCommandList->Close());
+
+    // Destroy the top layer and all its state.
+    _layerStack.pop();
```

Comment:
It once contained a unique_ptr so could not be copied; this made it a move.

I wanted to be able to move a popped value all at once, but thanks C++.

---
In reply to: [90589392](https://github.com/Microsoft/WinObjC/pull/1496#discussion_r90589392) [](ancestors = 90589392)